snowfall/training/ctc_graph.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c)  2020  Xiaomi Corp.       (author: Fangjun Kuang)
+
+from functools import lru_cache
+from typing import (
+    Iterable,
+    List,
+)
+
+import torch
+import k2
+import _k2
+
+
+def build_ctc_topo(phones: List[int]) -> k2.Fsa:
+    '''Build CTC topology.
+
+    The resulting topology converts repeated input
+    symbols to a single output symbol.
+
+    Args:
+      phones:
+        A list of phones
+    Returns:
+      Returns an FSA that converts repeated symbols to a single symbol.
+    '''
+    num_states = len(phones)
+    final_state = num_states
+    rules = ''
+    for i in range(num_states):
+        for j in range(num_states):
+            if i == j:
+                rules += f'{i} {i} {phones[i]} 0 0.0\n'
+            else:
+                rules += f'{i} {j} {phones[j]} {phones[j]} 0.0\n'
+        rules += f'{i} {final_state} -1 -1 0.0\n'
+    rules += f'{final_state}'
+    ans = k2.Fsa.from_str(rules)
+    return ans
+
+
+def compose(a_fsa: k2.Fsa, b_fsa: k2.Fsa) -> k2.Fsa:
```

Comment:
I was thinking we would write this function in `k2` instead of here.

Reply:
+1

Reply:
I wrote [one][1] some time ago in k2 but it was later removed.

https://github.com/k2-fsa/k2/pull/282#discussion_r513161115

https://github.com/k2-fsa/k2/pull/282#discussion_r513517896

[1]: https://github.com/k2-fsa/k2/commit/8538bd1e9a80e4a2835b4c189212bfc7a3e6e392

Reply:
I guess this can be moved to k2 for now.

Reply:
Removed to k2.